src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/ListViewModel.cs

```diff
@@ -31,80 +34,137 @@ public partial class ListViewModel : PageViewModel
     private void FetchItems()
     {
         // TEMPORARY: just plop all the items into a single group
         // see 9806fe5d8 for the last commit that had this with sections
-        // TODO unsafe
         try
         {
             var newItems = _model.Unsafe!.GetItems();
 
+            // Collect all the items into new viewmodels
+            Collection<ListItemViewModel> newViewModels = [];
+
+            // TODO we can probably further optimize this by also keeping a
+            // HashSet of every ExtensionObject we currently have, and only
+            // building new viewmodels for the ones we haven't already built.
             foreach (var item in newItems)
             {
-                // TODO: When we fetch next page of items or refreshed items, we may need to check if we have an existing ViewModel in the cache?
                 ListItemViewModel viewModel = new(item, this);
                 viewModel.InitializeProperties();
-                _itemCache.Add(viewModel); // TODO: Figure out when we clear/remove things from cache...
-
-                // We may already have items from the new items here.
-                if ((Filter == string.Empty || viewModel.MatchesFilter(Filter))
-                    && !Items.Contains(viewModel)) //// TODO: We should be smarter about the contains here somehow (also in OnFilterUpdated)
-                {
-                    // Am I really allowed to modify that observable collection on a BG
-                    // thread and have it just work in the UI??
-                    Items.Add(viewModel);
-                }
+                newViewModels.Add(viewModel);
             }
+
+            // Now that we have new ViewModels for everything from the
+            // extension, smartly update our list of VMs
+            ListHelpers.InPlaceUpdateList(Items, newViewModels);
+
+            // TODO: Iterate over everything in Items, and prune items from the
+            // cache if we don't need them anymore
         }
         catch (Exception ex)
         {
             ShowException(ex);
             throw;
         }
+
+        // Now that our Items contains everything we want, it's time for us to
+        // re-evaluate our Filter on those items.
+        if (!_isDynamic)
+        {
+            // A static list? Great! Just run the filter.
+            ApplyFilter();
+        }
+        else
+        {
+            // A dynamic list? Even better! Just stick everything into
+            // FilteredItems. The extension already did any filtering it cared about.
+            ListHelpers.InPlaceUpdateList(FilteredItems, Items);
+        }
+    }
+
+    /// <summary>
+    /// Apply our current filter text to the list of items, and update
+    /// FilteredItems to match the results.
+    /// </summary>
+    private void ApplyFilter()
+    {
+        Collection<ListItemViewModel> filtered = [.. FilterList(Items, Filter).ToList()];
+        ListHelpers.InPlaceUpdateList(FilteredItems, filtered);
```

Comment:
Won't the `ToList` be expensive here? Should we modify `InPlaceUpdateList`' to accept an enumerable for the new contents?

Reply:
Yea `IEnumerable` probably does make more sense - `InPlaceUpdateList` was originally ported from C++ and I was much worse at C# when I did that

\> implying I'm any good at c# now